src/ripple/app/misc/detail/WorkFile.h

```diff
@@ -0,0 +1,144 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2016 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_APP_MISC_DETAIL_WORKFILE_H_INCLUDED
+#define RIPPLE_APP_MISC_DETAIL_WORKFILE_H_INCLUDED
+
+#include <ripple/app/misc/detail/Work.h>
+#include <boost/system/error_code.hpp>
+#include <cassert>
+
+namespace ripple {
+
+namespace detail {
+
+// Work with files
+class WorkFile: public Work
+    , public std::enable_shared_from_this<WorkFile>
+{
+protected:
+    using error_code = boost::system::error_code;
+    // Override the definition in Work.h
+    using response_type = std::string;
+
+public:
+    using callback_type =
+        std::function<void(error_code const&, response_type&&)>;
+public:
+    WorkFile(
+        std::string const& path,
+        boost::asio::io_service& ios, callback_type cb);
+    ~WorkFile();
+
+    void run() override;
+
+    void cancel() override;
+
+private:
+    std::string path_;
+    callback_type cb_;
+    boost::asio::io_service& ios_;
+    boost::asio::io_service::strand strand_;
+
+};
+
+//------------------------------------------------------------------------------
+
+WorkFile::WorkFile(
+    std::string const& path,
+    boost::asio::io_service& ios, callback_type cb)
+    : path_(path)
+    , cb_(std::move(cb))
+    , ios_(ios)
+    , strand_(ios)
+{
+}
+
+WorkFile::~WorkFile()
+{
+    if (cb_)
+        cb_ (make_error_code(boost::system::errc::interrupted),
+            {});
+}
+
+void
+WorkFile::run()
```

Comment:
It might be a good idea to check the file size before attempting to load it into a string. As is, we will consume all memory with a large enough file. No change required but we may want to use a sensible threshold and log an error.

Reply:
On the one hand, yes, that's a risk. But on the other hand, this is a local file, loaded from local configuration (which is also read as a string), set up by the node operator. It's very unlikely we'd have to worry about malicious input here. Accidents I could see, but that seems really unlikely. So I don't really think a check is necessary. It wouldn't be too hard to implement with a max size of about 1Mb, though. What do you think?

Reply:
Considering that this is a local file, I'm not sure that really is a big consideration. I'm not opposed to adding this, but I don't think it's that's necessary either.

Reply:
@ximinez I agree with your comments, but also simple enough to add. It's your call.  `if (space(path_).available < 1024 * 1024)`

Reply:
Aren't there some custom functions now to calculate the number of bytes in a Megabyte?

I'd prefer adding this in fwiw.

Reply:
I pushed a commit which adds an optional size limit and threw in some tests while I was at it.

Reply:
@MarkusTeufelberger I vaguely remember those being added not too long ago or at least being discussed.